src/hotspot/share/tsan/tsanOopMapTable.cpp

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright (c) 2024, Google and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "tsan/tsanExternalDecls.hpp"
+#include "tsan/tsanOopMap.hpp"
+#include "tsan/tsanOopMapTable.hpp"
+
+TsanOopMapTableKey::TsanOopMapTableKey(oop obj) {
+  _wh = WeakHandle(TsanOopMap::oop_storage(), obj);
+  _obj = obj;
+}
+
+TsanOopMapTableKey::TsanOopMapTableKey(const TsanOopMapTableKey& src) {
+  _wh = src._wh;
+  _obj = src._obj;
+}
+
+void TsanOopMapTableKey::release_weak_handle() const {
+  _wh.release(TsanOopMap::oop_storage());
+}
+
+oop TsanOopMapTableKey::object_no_keepalive() const {
+  return _wh.peek();
+}
+
+void TsanOopMapTableKey::update_obj() {
+  oop obj = _wh.peek();
```

Comment:
`oopDesc*` is `oop`.

Reply:
`oopDesc*` is different from `oop` in a fastdebug build, when `CHECK_UNHANDLED_OOPS` is defined.

Reply:
See the `oop` class definition [here](https://github.com/openjdk/jdk/blob/0898ab7f7496689e5de52a5dc4530ca21def1fca/src/hotspot/share/oops/oopsHierarchy.hpp#L53-L115). Note that `CHECK_UNHANDLED_OOPS` is only enabled in [fastdebug](https://github.com/openjdk/jdk/blob/0898ab7f7496689e5de52a5dc4530ca21def1fca/make/hotspot/lib/JvmFlags.gmk#L77-L81), not slowdebug.
As said in the other comment,  we could declare `TsanOopMapTableKey::_obj` as an `oop`, but make sure to test with fastdebug build.

Reply:
I just reread your earlier suggestion more closely. 

> is it better to do oopDesc* obj = _wh.peek()?

WeakHandle only defines `peek()` with `oop` as return type (see `inline oop peek() const;`). With `CHECK_UNHANDLED_OOPS`, the implicit cast from `oop` to `oopDesc*` is handled by the `oopDesc* ()` operator. I think we should keep `oop obj = _wh.peek();` for consistency with the return type declared in `peek`. 

For the issue with `obj != _obj`, changing `obj != cast_to_oop(_obj)` resolves it.

Reply:
Yes, I'm aware `oopDesc* obj = _wh.peek()` will do the cast with the `oopDesc* ()` operator under `CHECK_UNHANDLED_OOPS`.

The point is to avoid unnecessary type casting and conversion if possible.  `oopDesc* obj = _wh.peek()` will just do the cast once, and avoid later casting with `obj != _obj` and `_obj = obj`.

Alternative is to declare `TsanOopMapTableKey::_obj` as an `oop`, which is probably cleaner.

Reply:
> For the issue with obj != _obj, changing obj != cast_to_oop(_obj) resolves it.

`cast_to_oop(_obj)` is not needed. This is addressed by changing `_obj` to `oop`.